lib/cretonne/src/simple_gvn.rs

```diff
@@ -0,0 +1,133 @@
+//! A simple GVN pass.
+
+use flowgraph::{ControlFlowGraph};
+use dominator_tree::DominatorTree;
+use ir::{Cursor, DataFlowGraph, InstructionData, Function, Inst, Opcode};
+use std::hash::{Hash, Hasher};
+use std::collections::HashMap;
+
+/// Test whether the given opcode is unsafe to even consider for GVN.
+fn trivially_unsafe_for_gvn(opcode: Opcode) -> bool {
+    opcode.is_call() || opcode.is_branch() || opcode.is_terminator() ||
+    opcode.is_return() || opcode.can_trap()
+}
+
+#[derive(Debug)]
+struct VisibleInst(InstructionData);
+
+impl Hash for VisibleInst {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // TODO: Is a better way to do this? Should we just make InstructionData
+        // implement Hash itself?
```

Comment:
You can `#[derive(Hash)]` too.

The problem is those instructions that store their arguments in a `ValueList`. They will hash the list reference, not the contents. It's possible you don't care about any of those instructions: jumps, calls, return.